### PR TITLE
Ipfs 0.4.0 changelog

### PR DESCRIPTION
**Pull request by whyrusleeping**
This PR merging will be the 0.4.0 release. See https://github.com/ipfs/go-ipfs/issues/2334#issuecomment-195046511 for a list of things we still need to do before merging this and shipping.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

**Comment by dignifiedquire** [comment]
Upgrade webui (when we actually have done it)

**Comment by jbenet** [comment]
@whyrusleeping @dignifiedquire where can i see the test runs' outputs of the test suites we discussed? This is the biggest pre-req to shipping 0.4.0. need to make sure everything is rock solid. (i know there's been a lot of activity on this, just point me to the things i should look at, i cant dig through lots of notifs yet)

**Comment by whyrusleeping** [comment]
@jbenet what sort of output do you want to see? A long list of what operations were executed?

**Comment by jbenet** [comment]
@whyrusleeping output similar to CI output, something that expresses what sort of operations were run, whether any problems happened and so on. Right now i have zero information on what the added testing does. Please also point me to PRs / repos i should look at and study.

**Comment by whyrusleeping** [comment]
@jbenet http://v04x.ipfs.io/ipfs/QmaFq8EBWLrJWCbH1xNBf7ywiqbK49dzrPLwsC3XPEyZPd

caution: 40MB log file

**Comment by whyrusleeping** [comment]
Also, when printing out what operations happen, it prints out the contents of any files or objects its creating, so you'll have to skim through that... I can try and silence that part of the logging

**Comment by whyrusleeping** [comment]
@jbenet better readable logs here: https://ipfs.io/ipfs/QmPMokr9nir533yosvKTgd64aQXjEwF1Ts78n1VJUSPndd

**Comment by dignifiedquire** [comment]
It would be great if these logs had references to commits and/or issues.

**Comment by jbenet** [comment]
Please also point me at things I should look at (the relevant PRs, randor, whatever), and how I can run the test suite myself

**Comment by whyrusleeping** [comment]
@dignifiedquire what do you mean references to commits and/or issues? The linked test runs were done on this PR.

@jbenet 
if you want to run the tests, pull down https://github.com/dignifiedquire/randor, npm install then (in bash) do something like 
```
DEBUG=randor:*,-randor:generate IPFS_EXEC=$GOPATH/bin/ipfs bin/randor run -l 200 -p 10
```

There are more parameters in [the readme](https://github.com/dignifiedQuire/randor#options), but the default is to run all the different operations, listed here: https://github.com/dignifiedquire/randor/tree/master/lib/operations

Note: each of those 'operations' is made up of a series of individual ipfs api calls

the 'issue' i mentioned it finding a week or so ago was a bug in randor calling pin rm twice on the same hash accidentally

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 28b32df9673cba1801b8a1665c2ffaf9389dd0ff
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by whyrusleeping** [comment]
@GitCop youre wrong. leave me alone.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 74767b19ca7b4fdb0ad91f3d6bc994fecc7b3688
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by lgierth** [comment]
Can I still add things? I'll have brief notes about X-Ipfs-Gateway-Prefix and the docker image.

**Comment by whyrusleeping** [comment]
oh, sure. go ahead

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 74767b19ca7b4fdb0ad91f3d6bc994fecc7b3688
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 74767b19ca7b4fdb0ad91f3d6bc994fecc7b3688
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by RichardLitt** [comment]
Ugh, @whyrusleeping, what's up with gitcop?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a7f469eb36cd0b0126e6ad5d33487e2113d93aec
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by whyrusleeping** [comment]
gitcop has no idea what its doing. its gone off the deep and and we need to put it down

**Comment by whyrusleeping** [comment]
we now use [gx](https://github.com/whyrusleeping/gx) to manage ipfs's dependencies. This means that you can no longer install ipfs with `go get`, instead run `make install` from the repo root.

**Comment by lgierth** [comment]
@whyrusleeping wdyt?

    * go-ipfs now uses [gx](https://github.com/whyrusleeping/gx) to manage its
      dependencies. This means that under the hood, go-ipfs's dependencies are
      backed by IPFS itself! It also means that go-ipfs is no longer installed
      using `go get`. Use `make install` instead.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0403306034886eb898c94656b0bf3c8072a19833
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,12 +1,133 @@
...
+  * Remove mentions of `/dns` from `ipfs dns` command docs. (@lgierth)
+* Testing
+  * Refactor iptb tests. (@chriscool)
+  * Improve t0240 sharness test. (@chriscool)
+  * Make bitswap tests less flaky. (@whyrusleeping)
+  * Use TCP port zero for ipfs daemon in sharness tests. (@whyrusleeping)
+  * Improve sharness tests on AppVeyor. (@chriscool)
+  * Add a pause to fix timing on t0065. (@whyrusleeping)
+  * Add support for arbitrary TCP ports to t0060-daemon.sh. (@noffle)
+  * Make t0060 sharness test use TCP port zero. (@whyrusleeping)
```
Comment (jbenet) [inline CHANGELOG.md]: - pinning stress tests?  
- randor stress tests?
- fs-repo-migration stress tests?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0403306034886eb898c94656b0bf3c8072a19833
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 7263c5c3141ec52fbc8eeefc4e55f75d17304ba4
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by jbenet** [comment]
@whyrusleeping merged this by hand.

**Comment by jbenet** [comment]
thanks everyone!

**Pull request closed without merging**